network/src/noise/error.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use libra_crypto::noise::NoiseError;
+use libra_types::PeerId;
+use short_hex_str::ShortHexStr;
+use std::io;
+use thiserror::Error;
+
+/// Different errors than can be raised when negotiating a Noise handshake.
+#[derive(Debug, Error)]
+pub enum NoiseHandshakeError {
+    #[error("noise client: missing remote server's public key when dialing")]
+    MissingRemotePublicKey,
+
+    #[error("noise client: error building handshake init message: {0}")]
+    BuildClientHandshakeMessage(NoiseError),
+
+    #[error("noise client: error sending client handshake init message: {0}")]
+    ClientWrite(io::Error),
+
+    #[error(
+        "noise client: error reading server handshake response message, server \
+         probably rejected our handshake message: {0}"
+    )]
+    ClientRead(io::Error),
+
+    #[error("noise client: error finalizing secure connection: {0}")]
+    ClientFinalize(NoiseError),
+
+    #[error("noise server: error reading client handshake init message: {0}")]
+    ServerRead(io::Error),
+
+    #[error("noise server: client peer id is malformed somehow: {0}")]
+    InvalidClientPeerId(String),
+
+    #[error("noise server: detected self-dial: we're trying to connect to ourselves")]
+    SelfDial,
+
+    #[error("noise server: client {0}: client is expecting us to have the wrong public key: {1}")]
+    ClientExpectingWrongPubkey(ShortHexStr, String),
+
+    #[error("noise server: client {0}: error parsing handshake init message: {1}")]
+    ServerParseClient(ShortHexStr, NoiseError),
+
+    #[error("noise server: client {0}: lock is poisoned")]
+    LockPoisoned(ShortHexStr),
+
+    #[error(
+        "noise server: client {0}: known client connecting to us with \
+         unauthenticated public key: {1}"
+    )]
```

Comment:
this means a client is connecting with a self-reported peer id I recognize (e.g., it's one of the validators' peer ids) but the public key they provided is not one of the public keys that we trust for that peer id. this could be a transient error during a network key rotation or someone attempting to impersonate a validator for example.